[2단계 - 로또 구현] 라젤(문선영) 미션 제출합니다.

RaZel713 opened the pull request:
얼렁뚱땅 우테코 LEVEL1. 로또  미션 구현 2단계
---

안녕하세요! 조앤 리뷰어님! 
1단계 미션에서 실수가 많았었는데 친절히 알려주셔서 정말 감사합니다!
1단계에서 이미 많은 부분 리팩터링을 해버려서... 남은 부분과 2차 피드백 반영하여 제출해 봅니다.

---

### 고려한 점

- 난수 생성: 언제나 일정한 확률을 가지고 난수가 생성될 수 있게 해야 한다. ('로또'이므로 공정성이 중요하다고 생각)
  - 처음에 random을 사용하였다가 SecureRandom로 바꾸었고 최종적으로 Collections.shuffle() 을 사용하였습니다!
  - 시간 복잡도와 복잡한 코드를 지양하고자 했습니다.

- 당첨 여부 확인: 긴 코드보다 간단하고 확실하게 뜻이 드러나는 코드를 작성하고자 했습니다!
  - set과 retainAll을 활용하여 작성했습니다.
 
---

### 리팩터링 point!

- test코드 관련 피드백을 반영하였습니다!

- README.md에 패키지 구조도를 추가하였습니다.

- Common-Utils-ValidationUtils: 유효성 검사에서 범용적으로 사용되는 메서드를 관리하였습니다! (check~: 유효성 검사에서 사용되는 범용적인 메서드 / validate~: 특정 유효성 검사 메서드)

- 공통적으로 사용되는 부분을 제네릭을 활용하여 표현하였습니다.

- View의 Prompt 관련 DisplayConstants를 추가하였습니다.

---

### 질문 point!

- Q1. README.md에 이해를 돕기 위해서 패키지 구조도를 추가하였는데, 적절한지 궁금합니다! 
패키지 구조도 때문에 너무 길어지면 읽기 번거로울 수도 있다는 생각이 들어서요. 
리뷰어로써는 어떤 점이 더 나으실 지 여쭤보고 싶습니다! (요약된 README vs 세세한 README)

- Q2. 페어가 도메인 속에 검증 로직을 넣는 것이 캡슐화가 잘 되는 것이라고 이야기하였는데 
저는 Validator를 작성하여 검증 로직을 Validator로 부터 불러오는 것이 '유효성 검사'를 관리하는 데 더 효과적이라고 생각했습니다. 
그래서 여러 번 사용하는 로직만 validator에 작성하기로 합의하였는데 리뷰어님께서는 어떻게 생각하실까요?

- Q3. inputView를 통해서 사용자에게 입력을 받고 해당 입력을 모델에 적용할 수 있는 최소한의 parsing을 하는 것을 service에 두었는데, 이렇게 작성해도 괜찮은지 궁금합니다. 
리팩터링 조건 사항에서 핵심 비지니스 로직을 가지는 객체를 model(도메인)에 두라고 이야기하였는데, 파싱 또한 model에 두는 것이 맞는지 의문이 들어서 질문드립니다!

RaZel713 committed:
refactor: test 코드 수정

- PrizeResultTest 매직넘버 21 제거
- setup의 result 부분 정리

RaZel713 committed:
refactor: AppConfig 조건문 수정

RaZel713 committed:
refactor: 테스트 코드 패키지 구조 변경

RaZel713 committed:
refactor(view): 뷰 관련 상수 enum 생성

seovalue commented on src/main/java/view/DisplayConstants.java:
```diff
@@ -0,0 +1,33 @@
+package view;
+
+public enum DisplayConstants {
```
👍 View용 Constants를 분리해주셨군요

seovalue commented:
라젤, 아시다시피 지난번 미션에서 step2까지 한꺼번에 진행해주셔서 이번 PR에는 나타나지 않는 변경사항들이 있는데요! 보기 힘들지만 그것 관련된 코멘트는 1단계 File changed 부분에 남겨둘테니 함께 확인하고 step2 브랜치에서 리팩터링에 반영해 리뷰 요청 주세요!!
코멘트에 대한 답변도 여기다 남겨주셔도 상관없어요!
>>  [관련 리뷰 링크](https://github.com/woowacourse/java-lotto/pull/576/files) 여기서 한번에 보셔도 되고, 아래에서 하나씩 보셔도 돼요!

- [README에 패키지 구조도를 나타내셨다고 했는데, 안보여요!?](https://github.com/woowacourse/java-lotto/pull/576/files#r1957215870)
- [라젤이 생각하시는 Config class의 역할이 무엇인가요?](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216097)
- [Parser 객체들이 담기는 패키지명을 service.parser로 해주신 이유가 궁금해요!](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216251)
- [Controller 관련 수정 코멘트1](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216441)
- [Controller 관련 수정 코멘트2](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216798)
- [Discriminator 관련 코멘트](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216577)
- [OwnedLotto 관련 코멘트1](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216930)
- [OwnedLotto 관련 코멘트2](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216947)
- [RandomNumberGenerator](https://github.com/woowacourse/java-lotto/pull/576/files#r1957217380)
- [Validator와 ValidationUtils](https://github.com/woowacourse/java-lotto/pull/576/files#r1957217561)

seovalue reviewed (changes requested):
라젤! 즐거운 일요일이네요ㅎㅎ
아시다시피 step1에서 2단계와 관련된 변경도 작업해주셨어서, 이번 PR에서는 변경사항이 제대로 나타나지 않았어요. 그래서 우선 관련 코멘트들은 (https://github.com/woowacourse/java-lotto/pull/595#issuecomment-2661172144) 여기에 한번에 모아서 남겨뒀으니, 모두 확인해보시고 이번 PR에 반영해서 리뷰 요청 주세요!!
리팩터링 하느라 고생많으셨어요~~ 👍

seovalue commented:
라젤, 아시다시피 지난번 미션에서 step2까지 한꺼번에 진행해주셔서 이번 PR에는 나타나지 않는 변경사항들이 있는데요! 보기 힘들지만 그것 관련된 코멘트는 1단계 File changed 부분에 남겨둘테니 함께 확인하고 step2 브랜치에서 리팩터링에 반영해 리뷰 요청 주세요!!
코멘트에 대한 답변도 여기다 남겨주셔도 상관없어요!
>>  [관련 리뷰 링크](https://github.com/woowacourse/java-lotto/pull/576/files) 여기서 한번에 보셔도 되고, 아래에서 하나씩 보셔도 돼요!

- [README에 패키지 구조도를 나타내셨다고 했는데, 안보여요!?](https://github.com/woowacourse/java-lotto/pull/576/files#r1957215870)
- [라젤이 생각하시는 Config class의 역할이 무엇인가요?](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216097)
- [Parser 객체들이 담기는 패키지명을 service.parser로 해주신 이유가 궁금해요!](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216251)
- [Controller 관련 수정 코멘트1](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216441)
- [Controller 관련 수정 코멘트2](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216798)
- [Discriminator 관련 코멘트](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216577)
- [OwnedLotto 관련 코멘트1](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216930)
- [OwnedLotto 관련 코멘트2](https://github.com/woowacourse/java-lotto/pull/576/files#r1957216947)
- [RandomNumberGenerator](https://github.com/woowacourse/java-lotto/pull/576/files#r1957217380)
- [Validator와 ValidationUtils](https://github.com/woowacourse/java-lotto/pull/576/files#r1957217561)

seovalue commented:
> Q1. README.md에 이해를 돕기 위해서 패키지 구조도를 추가하였는데, 적절한지 궁금합니다!
패키지 구조도 때문에 너무 길어지면 읽기 번거로울 수도 있다는 생각이 들어서요.
리뷰어로써는 어떤 점이 더 나으실 지 여쭤보고 싶습니다! (요약된 README vs 세세한 README)

저한테는 어쨌든 정보가 주어지는 면이니 세세한 README여도 좋아요! 세세한 README이면 어떤 이야기를 나눠야할 지 판단하기 좋을 것 같아요! (다만 제가 못찾은건지 위에도 남겼는데 패키지 구조도가 어딨나요..?!?!)
아니면 목적에 따라 파일을 구문해주셔도 좋아요!
- README에는 요구사항 위주
- PACKAGE.md에는 패키지 구조도 위주
이런 식으로요! 라젤 편하신대로 해주시면 될듯요~~


> Q2. 페어가 도메인 속에 검증 로직을 넣는 것이 캡슐화가 잘 되는 것이라고 이야기하였는데
저는 Validator를 작성하여 검증 로직을 Validator로 부터 불러오는 것이 '유효성 검사'를 관리하는 데 더 효과적이라고 생각했습니다.
그래서 여러 번 사용하는 로직만 validator에 작성하기로 합의하였는데 리뷰어님께서는 어떻게 생각하실까요?

만약 새로운 개발자가 라젤의 코드로 개발을 이어 하게 되었다고 가정해봐요.
예를 들어 Student라는 도메인은 생성자로 name, age를 받는데 name은 최대 3글자, age는 20살 미만만 받는다고 해볼게요.

새로운 개발자가 만약 Validator 없이 Student("안녕하세요", 100) 이렇게만 생성한다면, 정상적으로 생성될거예요. 
만약 도메인 속에 검증 로직을 넣었다면, Student("안녕하세요", 100) 이렇게 생성 시 예외가 발생하겠죠!!

어떻게 생각해요?


> Q3. inputView를 통해서 사용자에게 입력을 받고 해당 입력을 모델에 적용할 수 있는 최소한의 parsing을 하는 것을 service에 두었는데, 이렇게 작성해도 괜찮은지 궁금합니다.
리팩터링 조건 사항에서 핵심 비지니스 로직을 가지는 객체를 model(도메인)에 두라고 이야기하였는데, 파싱 또한 model에 두는 것이 맞는지 의문이 들어서 질문드립니다!

저도 이거 이야기해보고 싶었는데~ 여쭤봐주셔서 감사해요
우선 parsing이 service의 책임이라고 생각해주신 이유가 궁금하고, parsing의 역할은 무엇이며 어디에 영향을 미치는지 이야기해주실 수 있으실까요?!

RaZel713 commented:
꼼꼼하게 봐주셔서 감사합니다😊😊

Q2, 3은 관련 질문과 함께 답변 남깁니다!

Q1 답변: 목적에 따라 파일을 다르게 두는 것은 생각해 보지 못했는데! 좋은 방식 알아갑니다 

---

## 2단계 1차 피드백 반영 내역
---

1단계에도 해당 피드백에 대한 내용을 답변 달았으나 2단계에서 보시기 편하게 답변 이곳에 한번에 정리해 둡니다!

- **_FEEDBACK 1) README에 패키지 구조도를 나타내셨다고 했는데, 안보여요!?_**
답변: 패키지 구조도가 1단계 내용을 불러오는 과정에서 충돌로 날아간 것 같습니다! 
step1 -> step2로 넘어가는 과정에서 README.md에 충돌이 생겨서 백업해놓은 것을 가져다 썼는데 그게 패키지 구조도가 빠져있었습니다. 해당 부분 다시 반영했습니다!

---

- _**FEEDBACK 2) 라젤이 생각하시는 Config class의 역할이 무엇인가요?**_
답변: Config class의 역할은 객체의 생성 및 관리와 의존성 주입이라고 생각합니다.

---

- **_FEEDBACK 3) Parser 객체들이 담기는 패키지 명을 service.parser로 해주신 이유가 궁금해요!_**
**_(Q3. 우선 parsing이 service의 책임이라고 생각해주신 이유가 궁금하고, parsing의 역할은 무엇이며 어디에 영향을 미치는지 이야기해주실 수 있으실까요?!)_**
답변: 우선 Parser 객체들이 담기는 패키지명을 service.parser라고 둔 이유는 refactoring 전 코드에서는 service에서 3개의 parser 객체를 제외한 다른 객체(현재 model로 이동한)가 존재하였기 때문에 parser로 패키지를 만들어 3개의 parser들을 넣었습니다. 좀 더 보기 편하게 하기 위해서 parser 패키지를 만든 것인데 refactoring 과정에서 나머지 부분들은 model에 더 적합하다고 생각하여 이동하는 과정에 service패키지에 parser패키지만 남게 되었습니다.

   그리고 parsing의 역할에 대하여 답변 드리자면, parsing는 초반에 사용자에게 데이터를 입력 받고 해당 데이터의 입력이 입력 prompt (예를 들어, 보너스 번호를 입력해주세요.)에 맞는 내용인지 검증하고 model에 적용하기 위한 최소한의 처리를 하는데 사용하는 것으로 생각했습니다.

   프로그램의 핵심적인 로직을 처리하는 model에 해당 처리를 맡기는 것보다 service 단계를 만들어 model과 view 사이의 한 단계 완충 작용을 하는 것이 맡다고 생각하였습니다. 

---

- **_FEEDBACK 4)Controller 관련 수정 코멘트1: budget을 받아오는데 변수는 lottoCount인게 어색하게 느껴져요_**
답변: 처음 작성할 때에는 parser에 맞게 budget을 반환하게 작성하고,  int lottoCount = money / LOTTO_PRICE; 를 작성하여 Controller에서 로또의 개수를 구하는 방법으로 작성했었는데, lottoCount를 계산하는 로직이 짧고 간단하지만 Controller의 역할(model과 view를 연결하여 데이터를 전달)에는 맞지 않다고 생각하여 parser 단계로 넣었었는데, 해당 코드를 수정할 때 네이밍이 적합하지 못했던 것 같습니다. 해당 부분 수정하겠습니다! 

---

- **_FEEDBACK 5) Controller 관련 수정 코멘트2: return 0;과 return null;은 어떤 기준으로 구분해주신거예요?_**
답변: System.exit(1)를 사용하여 예외가 발생하면 프로그램이 종료되게 작성하였는데, 
해당 메서드에 return이 없으면 컴파일 오류가 발생하기 때문에 System.exit(1) 뒤에 return을 작성하였습니다. (System.exit(1);이 실행되면 프로그램이 종료되므로, return 이 실행될 가능성이 없다고 생각했습니다.)

   이때 중복되는 코드를 줄이기 위해 제네릭을 활용하여 코드를 작성하였는데, 
inputResponseForBonusNumber는 int 타입을 반환해야하기 때문에 기본 값(0)을 반환하게 작성하였고,
inputResponse는 제네릭 타입이기 때문에 null을 반환하게 작성하였습니다.
   -> 해당 부분 생각해보니 제네릭 타입이 기본형 int, double을 반환하는 경우는 null을 넣을 수 없기 때문에 Optional<T>를 사용하는 방법으로 수정하겠습니다!

---

- **_FEEDBACK 6) Discriminator 관련 코멘트: 객체명은 대문자로 시작하도록 수정할까요? 그리고 discriminator는 너무 범용적인 이름이라 정확히 어떤 역할을 하는지 알기가 어려운 것 같아요. 약간 '라젤'이여서 라젤을 구분할 수 있지, '사람'이라고 하게 되면 정확히 누구인지 알기 어려운 것처럼요! 내부를 메서드를 분리하는 등 가독성 높게 개선해보면 좋을 것 같아요!_**

  답변: 헉 소문자로 작성했었네요 수정하겠습니다! WinningDiscriminator 로 당첨 여부 판별을 하는 이름으로 작성하겠습니다.

---

- **_FEEDBACK 7) OwnedLotto 관련 코멘트1: Lotto가 아닌 OwnedLotto라고 명명해주신 이유가 궁금해요. Owned라는 건 소유자가 있다는 뜻인데, 객체만 봤을 때는 소유자가 드러나진 않는 것 같아서요. 그리고 Lotto.java 객체와 어떤 차이가 있는건지 잘 모르겠어요_**

  답변: Lotto객체는 한 장의 로또를 표현하기 위해서 작성하였습니다! (찾아보니 보통 로또가 5000원으로 A,B,C,D,E 이렇게 5줄 나오는 형식으로 판매가 되더라고요. 현재 코드는 1000원에 한 장, 한 줄 느낌이지만 확장성을 고려했을 때 로또 한 장을 표현하는 객체가 필요할 것으로 생각했습니다)

     Lottos에서 OwnedLotto라고 명명한 이유는 구매한 사람이 소유한 로또라고 표현하고자 해당 이름으로 작성했습니다. (한 장을 구매하는 경우 Lottos보다는 OwnedLotto가 낫다고 생각했습니다.)
   -> 생각해보니 구매한 로또가 더 나을 것 같아 purchasedLotto로 변경합니다.

---

- **_FEEDBACK 8) OwnedLotto 관련 코멘트2: index로 찾아온다고 하면, 외부에서는 해당 메서드를 호출할 때 무슨 인덱스로 호출하는 건지 직관적으로 알 수 있을까요?_**
  답변:  구매한 로또들이 생성 순서대로 저장되니 가장 처음에 존재하는 로또가 가장 먼저 생성된 로또이니,
  findLottoByCreationOrder 라고 네이밍하면 될까요?

  이 뜻으로 이야기하신게 맞으신가요..?

---

- **_FEEDBACK 9) RandomNumberGenerator:  .toList(); 그냥 이렇게도 가능해요_**
답변: 훨씬 간단한 것 같습니다! 해당 방법으로 변경하겠습니다

---

- _**FEEDBACK 10) Validator와 ValidationUtils: Validator와 ValidationUtils가 두 개로 나뉘어진 기준이 뭐예요?**_
답변: ValidationUtils는 유효성 검증 단계에 있어 범용적으로 활용되는 부분을 두었습니다. 
(예를 들어 문자열, 정규 표현식, 에러 메시지를 받아서 해당 문자열을 검증하고 예외 발생 시 에러 메시지 던지는 로직,
해당 로직은 변수들만 바꾸면 여러 방식으로 사용할 수 있다. 숫자만 검증, 문자열만 검증, 특정 형식 검증 등)

   그리고 Validator에는 해당 프로그램에서 유효성 검사 중 반복적으로 사용되는 부분을 두었습니다.
(로또 프로그램의 경우 해당 번호가 1~45의 범위가 맞는지 검사하는 메서드가 당첨 번호 입력, 보너스 번호 입력 검증 부분등 여러군데에서 사용된다.)

   Q2 답변과 함께 말씀 드리자면 domain 내에 검증 로직을 사용하지 않는 것이 아닌 검증 로직을 사용하는 것인데 검증 메서드를 해당 객체에 작성하지 않고 Validator에 작성하는 것을 의미합니다!

   Student 객체에서 Validator.validateStudentName() 과 Validator.validateStudentAge()를 불러와서 사용하는 방식으로요!

   리뷰어님께서 말씀해주신 예시로 이야기를 해보자면
   ValidationUtils - 문자열의 길이를 검사하는 checkStringLength() , 문자열의 형식을 검사하는 checkStringForm()

   Validator - ValidationUtils의 checkStringLength(), checkStringForm()를 활용해 이름의 길이를 검사하는 validateStudentName()를 작성

   domain(or service) - Student 객체 내에 validateStudentName()을 사용하여 유효성 검사

   이런 느낌으로 평상시에 사용하고 있었는데, validate 메서드 자체가 domain에서 작성되야 할까요..?
해당 메서드가 다른 곳에도 사용할 수 있다고 판단되어서요.

RaZel713 commented:
추가로 comment 남깁니다! 제가 Validator에서 궁금한 부분은 
1. 위에 말씀드린 것처럼 3단계로 나누어서 작성해도 괜찮은지 
2. Validator를 사용한다면 모든 검증 로직을 Validator에 둬도 괜찮은지 (한번만 해당 검증 로직이 사용된다고 하더라도 나중에 코드를 작성할 때 해당 로직을 또 사용할 수 있기 때문에 Validator로 두는 것이 좋지 않을까요?) 입니다!

RaZel713 committed:
refactor: discriminator 네이밍 개선

- 역할을 명확하기 위해 당첨 판별기로 네이밍
- 앞글자 소문자를 대문자로 변경

RaZel713 committed:
refactor: OwnedLotto 네이밍 개선

- 해당 객체의 메서드 명 getLottoByIndex를 findLottoByCreationOrder로 변경
- OwnedLotto를 PurchasedLotto로 변경

RaZel713 committed:
refactor: Optional을 사용하여 입력 메서드 개선

- 제네릭 타입이 기본형 int, double을 반환하는 경우, null을 넣을 수 없기 때문에 Optional를 사용하는 방법으로 수정

RaZel713 committed:
refactor: 코드 정리

RaZel713 committed:
docs: 패키지 구조도 추가

RaZel713 committed:
refactor: 변수명 변경

- BudgetParser 관련 네이밍 개선

RaZel713 committed:
refactor: 당첨 여부 분석기 구조 변경

seovalue commented on PACKAGE.md:
```diff
@@ -0,0 +1,39 @@
+## 패키지 구조도
```
👍 
혹시 라젤이 패키지 구조도를 나타내려고 생각하시게 된 계기가 무엇인지 궁금해요!!
보기엔 좋은데 유지보수엔 어려울 수도 있을 것 같아서요~~

seovalue commented on src/main/java/model/PurchasedLotto.java:
```diff
@@ -3,14 +3,14 @@
 import model.lotto.Lotto;
 
-public class OwnedLotto {
+public class PurchasedLotto {
```
답변: Lotto객체는 한 장의 로또를 표현하기 위해서 작성하였습니다! (찾아보니 보통 로또가 5000원으로 A,B,C,D,E 이렇게 5줄 나오는 형식으로 판매가 되더라고요. 현재 코드는 1000원에 한 장, 한 줄 느낌이지만 확장성을 고려했을 때 로또 한 장을 표현하는 객체가 필요할 것으로 생각했습니다)

Lottos에서 OwnedLotto라고 명명한 이유는 구매한 사람이 소유한 로또라고 표현하고자 해당 이름으로 작성했습니다. (한 장을 구매하는 경우 Lottos보다는 OwnedLotto가 낫다고 생각했습니다.)
-> 생각해보니 구매한 로또가 더 나을 것 같아 purchasedLotto로 변경합니다.



---

로또 1000원 단위로 판매돼요!!
5000원은 5000원치 구매해서 그런거예요

라젤이 어떤 의미로 구현해주신 건지 이해했어요.
예를 들어 로또 5천원치를 구매했다고 하면, 7개 번호로 구성된 `Lotto`가 총 5장 생기는거고, 이 5장을 `PurchasedLotto` 라고 명명해주신거군요!
여전히 어색하게 느껴지는 점은 사실 PurchasedLotto 안에서 Purchased를 나타내는 부분이 없기 때문에 어색하게 느껴지는 것 같아요. PurchasedLotto를 다른 개발자가 사용한다고 가정해봤을 때 purchasedLotto.purchasedAmount(), purchasedLotto.owner() 등등 `purchase` 와 관련있는 맥락들을 다 요청할 수 있게 느껴지는 것 같아요

seovalue commented on src/main/java/model/PurchasedLotto.java:
```diff
@@ -3,14 +3,14 @@
     }
 
-    public Lotto getLottoByIndex(int index) {
+    public Lotto findLottoByCreationOrder(int index) {
```
FEEDBACK 8) OwnedLotto 관련 코멘트2: index로 찾아온다고 하면, 외부에서는 해당 메서드를 호출할 때 무슨 인덱스로 호출하는 건지 직관적으로 알 수 있을까요?
답변: 구매한 로또들이 생성 순서대로 저장되니 가장 처음에 존재하는 로또가 가장 먼저 생성된 로또이니,
findLottoByCreationOrder 라고 네이밍하면 될까요?

이 뜻으로 이야기하신게 맞으신가요..?

---

네네, Index로 하게 되면 다음과 같은 질문들이 생길 수 있을 것 같아요!
그리고 또다른 의미는 만약 내부에서 순서가 보장되지 않는다면 어떤 일이 발생할 수 있을지?
외부에서는 idx를 어떻게 알 수 있을지? 
등등이 있을 것 같아요ㅎㅎ

작성해주신 `CreateOrder`가 Index 보다는 훨씬 명확한 것 같네요! 파라미터도 함께 변경되면 좋을 것 같아요!

seovalue commented on PACKAGE.md:
```diff
@@ -0,0 +1,39 @@
+    │       └───WinningNumberParser.java    // 당첨 번호 파서
+    ├───validator
+    │   ├───ErrorMessages.java              // 예외 메시지 enum
+    │   └───Validator.java                  // 유효성 검사기
```
Validator 관련 질문에 답변드려보자면

라젤이 의미해주신 바는 Validator는 좀 더 도메인에 강결합되어있는 검증 로직들이 분포하고, ValidateUtils는 좀 더 Common하게 사용될 수 있는 로직을 담고 있도록 구현해주셨다고 이해했어요.

저였다면
1. 도메인의 생성자에 해당 Validate 로직을 구현한다.
2. 생성자에 구현하는 input validation 로직이 중복되는 경우 (ex. String을 ","로 분리하는 로직이 여러 생성자에서 반복됨)
3. 해당 로직을 공통 로직으로 분리한다.
4. 도메인의 생성자에서 담당하는 Validator 로직이 점점 비대해진다면
5. 도메인 전용 Validator를 분리한다.

이런 식으로 작업할 것 같네요!

seovalue reviewed (approved):
안녕하세요 라젤!
남겨주신 질문에 대해 몇가지 코멘트 남겨뒀어요~ 다음 미션 진행하시기전에 확인해주세요.
첫번째 미션 하시느라 고생 많으셨어요~~
